engine/render/src/render/font_renderer.cpp

```diff
@@ -1047,15 +1048,23 @@ namespace dmRender
             case dmRender::RENDER_LIST_OPERATION_END:
                 {
                     uint32_t buffer_size = sizeof(GlyphVertex) * text_context.m_VertexIndex;
-                    dmGraphics::SetVertexBufferData(text_context.m_VertexBuffer, 0, 0, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
-                    dmGraphics::SetVertexBufferData(text_context.m_VertexBuffer, buffer_size, text_context.m_ClientBuffer, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
-                    text_context.m_VerticesFlushed = text_context.m_VertexIndex;
+                    dmhash_t hash = dmHashBuffer64(text_context.m_ClientBuffer, buffer_size);
+                    if (hash != text_context.m_PreviousHash)
+                    {
+                        DM_PROFILE(Render, "FontSetVertexBufferData");
+                        dmGraphics::SetVertexBufferData(text_context.m_VertexBuffer, 0, 0, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
+                        dmGraphics::SetVertexBufferData(text_context.m_VertexBuffer, buffer_size, text_context.m_ClientBuffer, dmGraphics::BUFFER_USAGE_STREAM_DRAW);
+                        text_context.m_VerticesFlushed = text_context.m_VertexIndex;
+                        text_context.m_PreviousHash = hash;
+                    }
                     DM_COUNTER("FontVertexBuffer", buffer_size);
                 }
                 break;
-            default:
-                assert(params.m_Operation == dmRender::RENDER_LIST_OPERATION_BATCH);
+            case dmRender::RENDER_LIST_OPERATION_BATCH:
                 CreateFontRenderBatch(render_context, params.m_Buf, params.m_Begin, params.m_End);
```

Comment:
Another (more complicated) optimization would be to avoid regenerating the local vertex buffer if a text entry has the same position and nothing in it has changed. See the "CreateFontRenderBatch" in the issue.